File: src/category.rs
use std::sync::Arc;

use super::error::AppError;
use super::DB_NAME;
use anyhow::anyhow;

use axum::routing::{delete, post};
use axum::{http::StatusCode, Router};
use axum_macros::debug_handler;
use serde::{Deserialize, Serialize};

use axum::extract::Query;

use mongodb::{bson::doc, Client};

use axum::extract::State;

pub(crate) const CATEGORY_COLLECTION: &str = "category";

#[derive(Serialize, Deserialize)]
pub(crate) struct Category {
    pub(crate) title: String,
    pub(crate) description: String,
    pub(crate) owner: String,
}

#[debug_handler]
pub(crate) async fn category(
    State(client): State<Arc<Client>>,
    Query(product): Query<Category>,
) -> Result<StatusCode, AppError> {
    let db = client.database(DB_NAME);

    let collection = db.collection::<Category>(CATEGORY_COLLECTION);

    collection
        .insert_one(product, None)
        .await
        .map_err(|e| anyhow!("Error: {e}"))?;

    Ok(StatusCode::OK)
}

#[derive(Serialize, Deserialize)]
pub(crate) struct CategoryId {
    category: String,
}

pub(crate) async fn delete_category(
    State(client): State<Arc<Client>>,
    Query(category): Query<CategoryId>,
) -> Result<StatusCode, AppError> {
    let collection = client
        .database(DB_NAME)
        .collection::<Category>(CATEGORY_COLLECTION);

    let filter = doc! {"title": category.category};

    collection.delete_one(filter, None).await?;

    Ok(StatusCode::OK)
}

pub(crate) fn get_category_routes(mongodb_client: Arc<Client>) -> Router {
    Router::new()
        .route("/", post(category))
        .route("/delete", delete(delete_category))
        .with_state(mongodb_client)
}

File: go_api/category.go
package main

import (
	"context"
	"net/http"
	"time"

	"github.com/labstack/echo/v4"
	"go.mongodb.org/mongo-driver/bson"
)

const category_collection = "category"

type Category struct {
	Title       string `query:"title"`
	Description string `query:"description"`
	Owner       string `query:"owner"`
}

func PostCategory(c echo.Context) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var category Category

	bind_err := (&echo.DefaultBinder{}).BindQueryParams(c, &category)
	if bind_err != nil {
		panic(bind_err)
	}

	collection := CLIENT.Database(DATABASE).Collection(category_collection)

	_, insert_err := collection.InsertOne(ctx, category)
	if insert_err != nil {
		return insert_err
	}

	return c.NoContent(http.StatusOK)
}

func DeleteCategory(c echo.Context) error {
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	var category Category

	bind_err := (&echo.DefaultBinder{}).BindQueryParams(c, &category)
	if bind_err != nil {
		panic(bind_err)
	}

	collection := CLIENT.Database(DATABASE).Collection(category_collection)

	_, insert_err := collection.DeleteOne(ctx, bson.M{"title": category.Title})
	if insert_err != nil {
		return insert_err
	}

	return c.NoContent(http.StatusOK)
}

func CreateCategorySubRoute(group *echo.Group) {
	group.POST("", PostCategory)
	group.DELETE("/delete", DeleteCategory)
}
